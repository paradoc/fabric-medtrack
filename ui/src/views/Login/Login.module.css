.login {
  display: grid;
  place-content: center;
  row-gap: 20px;
}

.form {
  display: grid;
  row-gap: 12px;
}

.form > input {
  height: 32px;
  width: 220px;
  border-radius: 8px;
  border: 1px solid lightgrey;
  padding: 4px 8px;
}

.form > input.-error {
  border: 1px solid lightcoral;
}

.form > button {
  height: 36px;
  width: 120px;
  border: none;
  border-radius: 8px;
  font-weight: bold;
  place-self: flex-end;
}

.form > button:enabled {
  background-color: aquamarine;
  color: #00774f;
}