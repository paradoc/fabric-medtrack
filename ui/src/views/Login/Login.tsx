import React, {
  ChangeEventHandler,
  KeyboardEvent,
  useCallback,
  useRef,
  useState,
} from 'react'
import { Navigate } from 'react-router-dom'
import { useSessionStorage } from 'usehooks-ts'
import { compose, values, includes } from 'rambda'

import styles from './Login.module.css'
import Logo from '../../components/Logo'

// In a real environment, this should be on backend.
const _CREDENTIALS: { [key: string]: string } = {
  medtrac: 'password',
}

export default function Login() {
  const [isLoggedIn, setIsLoggedIn] = useSessionStorage('login', false)
  const [formState, setFormState] = useState({
    username: false,
    password: false,
  })
  const [loginError, setLoginError] = useState(false)
  const usernameRef = useRef<HTMLInputElement>(null)
  const passwordRef = useRef<HTMLInputElement>(null)

  const submit = useCallback(() => {
    if (usernameRef.current && passwordRef.current) {
      const userVal = usernameRef.current.value
      const passVal = passwordRef.current.value
      const didError =
        !(userVal in _CREDENTIALS) || _CREDENTIALS[userVal] !== passVal
      setLoginError(didError)
      if (!didError) {
        setIsLoggedIn(true)
      }
    }
  }, [usernameRef, passwordRef])

  const checkNotEmpty = useCallback<ChangeEventHandler<HTMLInputElement>>(
    (e) => {
      const { alt } = e.currentTarget ?? {}
      if (alt) {
        setFormState((fs) => ({
          ...fs,
          [alt]: true,
        }))
      }
    },
    []
  )

  const handleKeyDown = useCallback(
    (e: KeyboardEvent<HTMLInputElement>) => {
      if (e.nativeEvent.isComposing || e.key !== 'Enter') return
      submit()
    },
    [submit]
  )

  return isLoggedIn ? (
    <Navigate to="/dispatcher" replace={true} />
  ) : (
    <div className={styles.login}>
      <Logo />
      <div className={styles.form}>
        <input
          type="text"
          placeholder="username"
          ref={usernameRef}
          onChange={checkNotEmpty}
          alt="username"
          className={loginError ? styles['-error'] : ''}
          onKeyDown={handleKeyDown}
          data-testid="user"
        />
        <input
          type="password"
          ref={passwordRef}
          onChange={checkNotEmpty}
          alt="password"
          className={loginError ? styles['-error'] : ''}
          onKeyDown={handleKeyDown}
          data-testid="pass"
        />
        <button
          disabled={compose(includes(false), values)(formState)}
          onClick={submit}
          data-testid="submit"
        >
          Login
        </button>
      </div>
    </div>
  )
}
