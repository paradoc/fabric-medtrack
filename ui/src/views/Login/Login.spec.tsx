import React from 'react'
import { render, fireEvent, waitFor, screen } from '@testing-library/react'
import Login from './Login'
import { act } from 'react-dom/test-utils'
import { MemoryRouter, Route, Routes } from 'react-router'

describe('Login', () => {
  test('should match snapshot', () => {
    const { container } = render(<Login />)
    expect(container).toMatchSnapshot()
  })

  test('should check behavior', async () => {
    const { container, getByTestId } = render(<Login />)
    const user = getByTestId('user')
    const pass = getByTestId('pass')
    const submit = getByTestId('submit')

    act(() => {
      fireEvent.change(user, { target: { value: 'test' } })
      fireEvent.change(pass, { target: { value: 'pass' } })
    })
    expect(submit).toBeEnabled()
    fireEvent.click(submit)
    expect(container).toMatchSnapshot()
  })

  test('should check behavior: ok', async () => {
    const { container, getByTestId } = render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route index element={<Login />} />
          <Route path="/dispatcher" element={<>dispatcher</>} />
        </Routes>
      </MemoryRouter>
    )
    const user = getByTestId('user')
    const pass = getByTestId('pass')
    const submit = getByTestId('submit')

    act(() => {
      fireEvent.change(user, { target: { value: 'medtrac' } })
      fireEvent.change(pass, { target: { value: 'password' } })
    })
    expect(submit).toBeEnabled()
    fireEvent.click(submit)
    expect(container).toMatchSnapshot()
  })
})
