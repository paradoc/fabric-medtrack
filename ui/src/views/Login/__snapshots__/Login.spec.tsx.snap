// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should check behavior 1`] = `
<div>
  <div
    class="login"
  >
    <div
      class="logo"
    >
      <strong>
        Mars
      </strong>
      <span>
        Drugstore
      </span>
    </div>
    <div
      class="form"
    >
      <input
        alt="username"
        class="-error"
        data-testid="user"
        placeholder="username"
        type="text"
      />
      <input
        alt="password"
        class="-error"
        data-testid="pass"
        type="password"
      />
      <button
        data-testid="submit"
      >
        Login
      </button>
    </div>
  </div>
</div>
`;

exports[`Login should check behavior: ok 1`] = `
<div>
  dispatcher
</div>
`;

exports[`Login should match snapshot 1`] = `
<div>
  <div
    class="login"
  >
    <div
      class="logo"
    >
      <strong>
        Mars
      </strong>
      <span>
        Drugstore
      </span>
    </div>
    <div
      class="form"
    >
      <input
        alt="username"
        class=""
        data-testid="user"
        placeholder="username"
        type="text"
      />
      <input
        alt="password"
        class=""
        data-testid="pass"
        type="password"
      />
      <button
        data-testid="submit"
        disabled=""
      >
        Login
      </button>
    </div>
  </div>
</div>
`;
